### PR TITLE
graphQl-961: ShippingAddressInput.postcode: String, is not required by Schema

### PR DESCRIPTION
**Pull request by kisroman**
### Fixed Issues (if relevant)

1. https://github.com/magento/graphql-ce/issues/961

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/QuoteGraphQl/Model/Cart/QuoteAddressFactory.php [hunk 1]
```diff
@@ -36,36 +38,55 @@ class QuoteAddressFactory
...
         BaseQuoteAddressFactory $quoteAddressFactory,
         GetCustomerAddress $getCustomerAddress,
-        AddressHelper $addressHelper
+        AddressHelper $addressHelper,
+        CountryInformationAcquirerInterface $countryInformationAcquirer = null
     ) {
         $this->quoteAddressFactory = $quoteAddressFactory;
         $this->getCustomerAddress = $getCustomerAddress;
         $this->addressHelper = $addressHelper;
+        $this->countryInformationAcquirer = $countryInformationAcquirer;
```
Comment (lenaorobei) [inline app/code/Magento/QuoteGraphQl/Model/Cart/QuoteAddressFactory.php:61]: We do not care about BC in GraphQL modules. Only the schema is out API.
https://github.com/magento/graphql-ce/wiki/Approach-to-Implementation#backward-compatibility-policy

File: app/code/Magento/QuoteGraphQl/Model/Cart/SetBillingAddressOnCart.php [hunk 2]
```diff
@@ -111,6 +111,45 @@ private function createBillingAddress(
...
+        }
+    }
+
+    /**
+     * Collecting errors.
+     *
+     * @param array $errors
+     * @return string
+     */
+    private function getAddressErrors(array $errors): string
```
Comment (lenaorobei) [inline app/code/Magento/QuoteGraphQl/Model/Cart/SetBillingAddressOnCart.php]: Please `@todo` with reference to https://github.com/magento/graphql-ce/issues/970

File: app/code/Magento/QuoteGraphQl/Model/Cart/SetShippingAddressesOnCart.php [hunk 3]
```diff
@@ -82,6 +83,44 @@ public function execute(ContextInterface $context, CartInterface $cart, array $s
...
+        }
+    }
+
+    /**
+     * Collecting errors.
+     *
+     * @param array $errors
+     * @return string
+     */
+    private function getAddressErrors(array $errors): string
```
Comment (lenaorobei) [inline app/code/Magento/QuoteGraphQl/Model/Cart/SetShippingAddressesOnCart.php]: Please add also test case with multiple errors.

**Review by lenaorobei — CHANGES_REQUESTED** [review]
Please address comments.

File: app/code/Magento/QuoteGraphQl/Model/Cart/QuoteAddressFactory.php (hunk 1 above)
Comment (kisroman) [reply app/code/Magento/QuoteGraphQl/Model/Cart/QuoteAddressFactory.php:61]: Ok.

File: app/code/Magento/QuoteGraphQl/Model/Cart/SetShippingAddressesOnCart.php (hunk 3 above)
Comment (kisroman) [reply app/code/Magento/QuoteGraphQl/Model/Cart/SetShippingAddressesOnCart.php]: Done.

File: app/code/Magento/QuoteGraphQl/Model/Cart/SetBillingAddressOnCart.php (hunk 2 above)
Comment (kisroman) [reply app/code/Magento/QuoteGraphQl/Model/Cart/SetBillingAddressOnCart.php]: Done.

**Comment by lenaorobei** [comment]
This PR has a dependency on https://github.com/magento/graphql-ce/issues/970

Moving to `on hold` to avoid double work with fixing error formatting.

**Comment by TomashKhamlai** [comment]
As a customer, I try to set shipping on empty cart
#### Request
```graphql
mutation {
  setShippingAddressesOnCart(
    input: {
      cart_id: "12345678901234567890123456789012"
      shipping_addresses: {
        address: {
          firstname: "John"
          lastname: "Doubledo"
	  telephone: ""
          save_in_address_book: true
          city: "Culver city"
        }
      }
    }
  ) {
    cart {
      email
      shipping_addresses {
        postcode
        firstname
        lastname
        telephone
        selected_shipping_method {
          carrier_code
          method_code
        }
        available_shipping_methods {
          carrier_code
          method_code
        }
      }
    }
  }
}
```

### Response
```graphql
{
  "errors": [
    {
      "debugMessage": "Notice: Undefined index: street in /var/www/html/graphqlce/app/code/Magento/QuoteGraphQl/Model/Cart/QuoteAddressFactory.php on line 89",
      "message": "Internal server error",
      "extensions": {
        "category": "internal"
      },
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "setShippingAddressesOnCart"
      ]
    }
  ],
  "data": {
    "setShippingAddressesOnCart": null
  }
}
```

**Comment by lenaorobei** [comment]
@TomashKhamlai query you are trying to test has syntax issues. `shipping_addresses` should be array.
```
shipping_addresses: [
        {
          address: {
            street: ["Streer"]
            country_code: "US"
            firstname: "John"
            lastname: "Doubledo"
            telephone: ""
            save_in_address_book: true
            city: "Culver city"
          }
        }
      ]
```

**Comment by lenaorobei** [comment]
@magento run Functional Tests

**Review by lenaorobei — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @lenaorobei, thank you for the review.
[ENGCOM-6170](https://jira.corp.magento.com/browse/ENGCOM-6170) has been created to process this Pull Request
:eight_spoked_asterisk: @lenaorobei, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by TomashKhamlai** [comment]
@lenaorobei, I've tried with array syntax also (usually both variants work as expected)

#### Request:
```graphql
mutation {
  setShippingAddressesOnCart(
    input: {
      cart_id: "02GYdVCktyqGo8LfEkoBwPbkCLxt3lL9"
      shipping_addresses: [
        {
          address: {
            country_code: "US"
            firstname: "John"
            lastname: "Doubledo"
            telephone: ""
            save_in_address_book: true
            city: "Culver city"
          }
        }
      ]
    }
  ) {
    cart {
      email
      shipping_addresses {
        postcode
        firstname
        lastname
        telephone
        selected_shipping_method {
          carrier_code
          method_code
        }
        available_shipping_methods {
          carrier_code
          method_code
        }
      }
    }
  }
}
```

#### Response:
```graphql
{
  "errors": [
    {
      "debugMessage": "Notice: Undefined index: street in /var/www/html/graphqlce/app/code/Magento/QuoteGraphQl/Model/Cart/QuoteAddressFactory.php on line 93",
      "message": "Internal server error",
      "extensions": {
        "category": "internal"
      },
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "setShippingAddressesOnCart"
      ]
    }
  ],
  "data": {
    "setShippingAddressesOnCart": null
  }
}
```

File: app/code/Magento/QuoteGraphQl/Model/Cart/QuoteAddressFactory.php [hunk 4]
```diff
@@ -36,36 +38,57 @@ class QuoteAddressFactory
...
+                throw new GraphQlInputException(__('The country isn\'t available.'));
+            }
+            $availableRegions = $countryInformation->getAvailableRegions();
+            if (null !== $availableRegions) {
+                $addressInput['region_code'] = $addressInput['region'];
+            }
+        }
+
         $maxAllowedLineCount = $this->addressHelper->getStreetLines();
         if (is_array($addressInput['street']) && count($addressInput['street']) > $maxAllowedLineCount) {
```
Comment (TomashKhamlai) [inline app/code/Magento/QuoteGraphQl/Model/Cart/QuoteAddressFactory.php:93]: I think that `$addressInput['street']` can be empty and this should be checked here or somewhere else before these lines can be executed.

**Comment by lenaorobei** [comment]
@magento run Unit Tests

**Comment by TomashKhamlai** [comment]
Created a separate issue #1033 for the case when the street parameter was omitted.

**Pull request merged**